docroot/sites/policy.uiowa.edu/modules/policy_core/src/Controller/PDFContentController.php

```diff
@@ -0,0 +1,331 @@
+        '#url' => Url::fromRoute('policy_core.pdf_content_download'),
+        '#attributes' => ['class' => ['button', 'button--primary']],
+      ];
+      $markup['regenerate'] = [
+        '#type' => 'link',
+        '#title' => $this->t('Regenerate PDF'),
+        '#url' => Url::fromRoute('policy_core.pdf_content', [], ['query' => ['regenerate' => 1]]),
+        '#attributes' => ['class' => ['button']],
+      ];
+      return $markup;
+    }
+
+    // If regenerating, delete the file.
+    if ($request->query->get('regenerate') && file_exists($real_path)) {
+      unlink($real_path);
+    }
+
+    // Start the batch process.
+    $batch_builder = new BatchBuilder();
+    // Starting menu link plugin ID.
+    // Format as `menu_link_content:{uuid}`.
+    $mpid = self::ROOT_MPID;
+    $nodes = self::getNodesByMenuOrder('main', 'page', $mpid);
+
+    $batch_size = 20;
+    foreach (array_chunk($nodes, $batch_size) as $chunk) {
+      $batch_builder->addOperation([static::class, 'batchProcess'], [$chunk]);
+    }
+
+    $batch_builder
+      ->setTitle($this->t('Generating Policy Manual PDF...'))
+      ->setInitMessage($this->t('Starting PDF generation...'))
+      ->setProgressMessage($this->t('Processed @current out of @total. Please wait while content is fetched and combined into a single PDF.'))
+      ->setErrorMessage($this->t('An error occurred during PDF generation.'))
+      ->setFinishCallback([static::class, 'batchFinished']);
+
+    batch_set($batch_builder->toArray());
+    return batch_process(Url::fromRoute('policy_core.pdf_content_finish')->toString());
+  }
+
+  /**
+   * Batch operation callback.
+   */
+  public static function batchProcess(array $nodes_chunk, array &$context): void {
+    $renderer = \Drupal::service('renderer');
+    $fs = \Drupal::service('file_system');
+
+    $html = '';
+    foreach ($nodes_chunk as $node) {
+      $render_array = \Drupal::entityTypeManager()
+        ->getViewBuilder('node')
+        ->view($node['node'], 'pdf');
+
+      $parent_title = $node['parent_title'];
+      if ($parent_title) {
+        $parent_markup = [
+          '#markup' => $parent_title,
+        ];
+        $render_array['parent_title'] = $parent_markup;
+      }
+
+      $html .= '<div class="pdf-page">' . $renderer->render($render_array) . '</div>';
+    }
+
+    $temp_dir = self::TEMP_DIR;
+    $fs->prepareDirectory($temp_dir, FileSystemInterface::CREATE_DIRECTORY);
+    $temp_file = tempnam($fs->realpath($temp_dir), 'batch_') . '.html';
+    file_put_contents($temp_file, $html);
+
+    $context['results']['temp_files'][] = $temp_file;
+    $context['message'] = t('Processed @count nodes in this batch.', ['@count' => count($nodes_chunk)]);
+  }
+
+  /**
+   * Batch finished callback.
+   */
+  public static function batchFinished($success, $results, $operations): void {
+    // Increase the memory limit temporarily for PDF generation.
+    ini_set('memory_limit', '512M');
```

Comment:
Will we need this for prod?  What is our current `memory_limit`?

Reply:
256M

Reply:
Good question. Ran into the issue in DEV (128). Maybe it won't be an issue in prod. I'd be willing to take this out until we know it is a problem there.